### PR TITLE
fix: reload once when log rotate

### PR DESCRIPTION
**Pull request by monkeyDluffy6017**
### Description

1.  Waiting for nginx reopen files to avoid losing logs during compression
2.  Refactor code to ensure that the signal is sent once at a time

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by monkeyDluffy6017** [comment]
> Is it feasible to add some test cases?
We have log-rotate.t, log-rotate2.t, log-rotate3.t  3 test files to test its functionality.
But I have no idea how to add test case about losing log file or sending several reopen signal.

File: apisix/plugins/log-rotate.lua [hunk 1]
```diff
@@ -37,7 +37,9 @@ local str_format = string.format
 local str_reverse = string.reverse
 local tab_insert = table.insert
 local tab_sort = table.sort
-
+local new_tab = require "table.new"
+local ngx_sleep = require("apisix.core.utils").sleep
+local isempty = require "table.isempty"
```
Comment (spacewander) [inline apisix/plugins/log-rotate.lua:42]: Better to introduce it via `core.table `

File: apisix/plugins/log-rotate.lua [hunk 2]
```diff
@@ -218,25 +221,42 @@ local function file_size(file)
...
+    core.log.warn("send USR1 signal to master process [", pid, "] for reopening log file")
+    local ok, err = signal.kill(pid, signal.signum("USR1"))
+    if not ok then
+        core.log.error("failed to send USR1 signal for reopening log file: ", err)
+    end
+
+    if enable_compression then
+        -- Waiting for nginx reopen files
+        -- to avoid losing logs during compression
+        ngx_sleep(wait_time)
```
Comment (spacewander) [inline apisix/plugins/log-rotate.lua:252]: Will the sleep block the timer and cause the next rotation time incorrect?

Comment (monkeyDluffy6017) [reply apisix/plugins/log-rotate.lua:252]: `wait_time` is just 0.5s by default. rotation time is much lager than 0.5s, right?

File: apisix/plugins/log-rotate.lua (hunk 1 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/log-rotate.lua]: Done

**Review by tzssangglass — DISMISSED** [review]

File: apisix/plugins/log-rotate.lua (hunk 2 above)
Comment (spacewander) [reply apisix/plugins/log-rotate.lua:248]: @monkeyDluffy6017 
Currently, there is no limitation for the user-specific wait_time. BTW, we need to doc it if this argument is exposed to the user.

Comment (spacewander) [reply apisix/plugins/log-rotate.lua:248]: If the wait time is only for hardcode, maybe we can add extra sleep in the test so we don't need to set a special value.

Comment (monkeyDluffy6017) [reply apisix/plugins/log-rotate.lua:248]: ok

Comment (monkeyDluffy6017) [reply apisix/plugins/log-rotate.lua]: Done

**Review by soulbird — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**